src/Meshes/thinInstanceMesh.ts

```diff
@@ -228,12 +238,34 @@ Mesh.prototype.thinInstanceBufferUpdated = function(kind: string): void {
     if (kind === "matrix") {
         if (this._thinInstanceDataStorage.matrixBuffer) {
             this._thinInstanceDataStorage.matrixBuffer!.updateDirectly(this._thinInstanceDataStorage.matrixData!, 0, this._thinInstanceDataStorage.instancesCount);
+            this._thinInstanceDataStorage.worldMatrices = null;
         }
     } else if (this._userThinInstanceBuffersStorage?.vertexBuffers[kind]) {
         this._userThinInstanceBuffersStorage.vertexBuffers[kind]!.updateDirectly(this._userThinInstanceBuffersStorage.data[kind], 0);
     }
 };
 
+Mesh.prototype.thinInstanceGetWorldMatrices = function(): Matrix[] {
+    if (!this._thinInstanceDataStorage.matrixData || !this._thinInstanceDataStorage.matrixBuffer) {
+        return [];
+    }
+    const matrixData = this._thinInstanceDataStorage.matrixData;
+
+    if (!this._thinInstanceDataStorage.worldMatrices || this._thinInstanceDataStorage.worldMatrices.length !== this._thinInstanceDataStorage.instancesCount) {
+        this._thinInstanceDataStorage.worldMatrices = new Array<Matrix>();
+
+        for (let i = 0; i < this._thinInstanceDataStorage.instancesCount; ++i) {
+            this._thinInstanceDataStorage.worldMatrices[i] = Matrix.Zero();
+        }
+    }
+
+    for (let i = 0; i < this._thinInstanceDataStorage.instancesCount; ++i) {
+        Matrix.FromArrayToRef(matrixData, i * 16, this._thinInstanceDataStorage.worldMatrices[i]);
+    }
+

```

Comment:
My bad in the comment above: you copy the matrices each time the function is called, so no sync problems with the thin instance matrices!

However, with the proposed changes on dirtyness, this could be added in the `if` block above and not done each time we call the function (with tens of thousands thin instances, it may make a difference (?)). But to be efficient, the previous array of world matrices should be kept somewhere before nulling it to avoid recreating the array + empty matrices in case the current thin instance matrix buffer length is the same than the length of the saved matrix array.

Reply:
Yeah It thought about it 

> Also, don't forget to update the doc: https://doc.babylonjs.com/how_to/how_to_use_thininstances ("Support" section) 😃

Yes as soon as it will be merged, I will add more doc for it